router/src/lib.rs

```diff
@@ -1,90 +1,109 @@
+        Self { addr, ctrl: tx }
     }
 
-    pub fn add_connection(&mut self, plug: Plug, routes: Vec<Ipv4Route>) {
-        let (tx, rx) = plug.split();
-        self.rxs.push(rx);
-        self.txs.push((tx, routes));
+    pub fn add_connection(&self, id: usize, plug: Plug, routes: Vec<Ipv4Route>) {
+        self.ctrl
+            .unbounded_send(RouterCtrl::AddRoute(id, plug, routes))
+            .ok();
     }
 
-    fn process_packet(&mut self, bytes: Vec<u8>) {
-        let packet = if let Some(packet) = Ipv4Packet::new(&bytes) {
-            packet
-        } else {
-            log::info!("router {}: dropping invalid ipv4 packet", self.addr);
-            return;
-        };
-        let dest = packet.get_destination();
-        if dest == self.addr {
-            log::info!("router {}: dropping packet addressed to me", self.addr);
-            return;
-        }
-        for (tx, routes) in &self.txs {
-            for route in routes {
-                if route.dest().contains(dest) || dest.is_broadcast() || dest.is_multicast() {
-                    log::debug!("router {}: routing packet on route {:?}", self.addr, route);
-                    let _ = tx.unbounded_send(bytes);
-                    return;
-                }
-            }
-        }
-        log::info!(
-            "router {}: dropping unroutable packet to {}",
-            self.addr,
-            dest
-        );
+    pub async fn remove_connection(&self, id: usize) -> Option<Plug> {
+        let (tx, rx) = oneshot::channel();
+        self.ctrl
+            .unbounded_send(RouterCtrl::RemoveRoute(id, tx))
+            .unwrap();
+        rx.await.unwrap()
     }
 }
 
-impl Future for Ipv4Router {
-    type Output = ();
-
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        let mut i = 0;
-        while i < self.rxs.len() {
-            loop {
-                let packet = match Pin::new(&mut self.rxs[i]).poll_next(cx) {
-                    Poll::Pending => {
-                        i += 1;
-                        break;
+fn router(addr: Ipv4Addr, mut ctrl: mpsc::UnboundedReceiver<RouterCtrl>) {
+    async_global_executor::spawn(async move {
+        let mut conns = vec![];
+        loop {
+            futures::select! {
+                ctrl = ctrl.next() => match ctrl {
+                    Some(RouterCtrl::AddRoute(id, plug, routes)) => {
+                        conns.push((id, plug, routes));
                     }
-                    Poll::Ready(None) => {
-                        self.rxs.remove(i);
-                        self.txs.remove(i);
-                        break;
+                    Some(RouterCtrl::RemoveRoute(id, ch)) => {
+                        let plug = if let Some(idx) = conns.iter().position(|(id2, _, _)| *id2 == id) {
+                            let (_, plug, _) = conns.swap_remove(idx);
+                            Some(plug)
+                        } else {
+                            None
```

Comment:
When does this return None?

Reply:
if there is no route to `id`, which could be for example because the machine exited.